### ISSUE TITLE
x10^ only accepts integers

### ISSUE THREAD

**Issue by artaxxx** [numworks/epsilon#124]
Type 4E3.5 or 4E(3.5)
The displayed result is 4000*0.5
It should be 4*10^(3.5) = 12649.11

**Comment by bernouilli** [comment]
I disagree:  4E3.5 should return an error
The notation xxxEyy is only valid if yy is in integer
I doubt there are other calculators that accept 4E3.5

**Comment by artaxxx** [comment]
You're right! This is definitely a syntax error

**Comment by Zezombye** [comment]
Meh, it shouldn't. Casio has 2 keys: the E that doesn't work with variables and decimals, and a small 10 that does work with variables+decimals. The E on the numworks should work with variables and decimals, there is no reason it shouldn't.

**Comment by ian-abbott** [comment]
Another oddity is that it allows 1.2.3 and treats it as an implicit multiplication 1.2*0.3 = 0.36 .

**Comment by adriweb** [comment]
Heh, I'm not really surprised, the multiplication is implicit, and in fact, I just checked, the TI-Nspire does the same.
That said, it's not everyday you'd type this :P

(Regarding EE and x10^, TI calcs also have two distinct keys for it, and with EE indeed you can't have decimals (on the Nspire for instance, `1.2e3.4` gets interpreted as `1.2e3*0.4`)

**Comment by Zezombye** [comment]
> Should implicit multiplication be allowed between two numbers?

I've never seen "1.2.3" written for expressing "1.2\*0.3", so no. (the whole .3 = 0.3 thing is dumb anyway)

**Comment by bernouilli** [comment]
The confusion comes from the fact that the numworks is missing the AntiLog key (10^x or ALOG).
Nevertheless, the x10^x key, although is labeled this way, should behave as the E ou EE key on other calculators

**Comment by adriweb** [comment]
For the x10^, allowing decimals isn't an issue on any other recent calcs having natural input, since you'll be typing the number (or anything else) in the superscript part dedicated for it

**Comment by ian-abbott** [comment]
Some calculators (for example HP 50g) do not allow implicit multiplication at all, so that's one way to fix it!

**Comment by Zezombye** [comment]
We need implicit multiplication, else it would be a regression (also, it should be in the output, currently it displays 5*i when it should display 5i).

The E should have a lower precedence than the decimal point, so that 5e3.5 is parsed as (5)e(3.5), because (5e3).5 is not what users expect.

**Comment by bernouilli** [comment]
the x10^ key is used to enter a number, not to perform a computation.
There is no cas where you would enter a number 2E3.5. In case you want to compute 2*ALOG(3.5), you wouldn't enter 2E3.5 but you would enter 2*10^3.5. 
The use of the key x10^ to compute the ALOG isn't good because x10^ implies that there is a number entered before the key which isn't necessary when you compute ALOG(something). 
As said before, the numworks is missing the ALOG key, this choice can be legitim since you can replace the ALOG function by simply using 10^X, but this shall not be done using the x10^ key.

**Comment by ian-abbott** [comment]
Perhaps we can just live with the anomaly since TI Nspire behaves the same way. I agree that the x10^x key shouldn't be used as an ALOG function; it's in the wrong section of the keyboard for a start. Perhaps the printing on the key could be changed to avoid it being confused with an ALOG function.

**Comment by ian-abbott** [comment]
I don't know enough about YACC, but could sequences such as `number number` and `number ICOMPLEX number` be made to trigger a syntax error in "poincare/src/expression_parser.y"? That would make entries such as 1.2.3, 2e3.5, and 4i5 return a syntax error but still allow implicit multiplication in general.

**Comment by adriweb** [comment]
Pretty much every language behaves the same way: "2e3.5" is a syntax error in Python, Lua, JS, C, C++...
It just happens that on the Nspire, the mathematical aspects of things made it so that it's parsing as an implicit multiplication (which also proves that the "e" operator also doesn't accept decimals).

The only exception I've seen so far is mathematica (but it's smart enough to infer the meaning of `e` in all cases you might throw at it)

**Comment by Zezombye** [comment]
>The only exception I've seen so far is mathematica (but it's smart enough to infer the meaning of e in all cases you might throw at it)

Well, shouldn't numworks be smart too? :p
Since the key is called *10^x (and not E) it should be a shortcut for writing *10^. So 3e3.5 is parsed like 3*10^3.5, which is 3*10^(3.5).
If you really think that we should keep the fact that the E operator is undefined for decimals, then make it so that pressing that key writes *10^ instead of E. Or make it like Casio, and write a subscript 10. But I find it dumb to throw a syntax error, instead of doing the expected behavior (which is 3*10^(3.5)).

**Comment by ian-abbott** [comment]
This is probably not quite right, but adding the following to the `exp:` production rule between the `| number` and `| ICOMPLEX` lines forces a syntax error for two consecutive numbers:

      | number number %prec IMPLICIT_MULTIPLY { Poincare::Expression * terms[2] = {$1,$2}; $$ = new Poincare::Multiplication(terms, false);  YYERROR; }

I basically copied the `| exp exp` part but added the YYERROR on the end. I'm not sure how much of the other junk is needed.

**Comment by Ecco** [comment]
See #151

**Comment by artaxxx** [comment]
I feel like I opened up a Pandora's box with this issue :) Thanks for commenting and making a great discussion with this!

To sum up everybody's point:

* E and 10^x keys have different behaviors: E only accepts integers by convention and 10^x is used for ALOG. So as far as the NumWorks calculator only got the E key, it should only be used with integers. ALOG is not on the keyboard because it's not that annoying to just enter 10^3.2 manually. I don't know what you guys think but maybe we can put an ALOG function in the toolbox menu. Personally I'm not comfortable with adding a shortcut for ALOG
* Remove implicit multiplication: when we were designing the interface we tried to remove implicit multiplication but, believe me, students found it was a pain to use 
* Implicit multiplication: I tested other calculators after your remarks, 1.2.3 = 1.2x0.3 is not really odd... some consider this computation as an error and others as 1.2x0.3. An error pop-up is always violent for the end user, so it's important to limit the use of modals or errors. In addition, as far as the expression is displayed in pretty print, it's easy to check how the expression is understood by the firmware. Then, when you compute 1E3.5 and get 1000E0.5 displayed, the result is pretty clear for the user.

In conclusion, the best is to leave things as they are. That's why I'm closing this issue. Thanks for your contribution!
